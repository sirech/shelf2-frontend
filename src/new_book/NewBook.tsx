import React from 'react'
import { useNavigate } from 'react-router-dom'
import { Field, Form, Formik, FormikProps } from 'formik'

import { Button, Container, Row, Col } from 'reactstrap'

import ReactstrapSelect from './ReactstrapSelect'
import input from './input'
import Stars from './Stars'

import initialBook from './initial_book'
import bookSchema from './schema'

import { useAppDispatch } from 'hooks'
import { actions } from 'state/form'

import type { BookForm, Categories } from 'types'

const categories: Categories[] = [
  'sociology',
  'software',
  'econ',
  'history',
  'other',
]

const NewBookWrapper = () => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()
  return (
    <Formik
      initialValues={initialBook}
      initialStatus={{ submitError: '' }}
      validationSchema={bookSchema}
      // eslint-disable-next-line @typescript-eslint/require-await
      onSubmit={async (values, { setStatus }) =>
        dispatch(
          actions.create(values, navigate, (error) =>
            setStatus({ submitError: error })
          )
        )
      }
    >
      {(props) => <NewBook {...props}></NewBook>}
    </Formik>
  )
}

const NewBook = ({
  status: { submitError },
  isSubmitting,
}: FormikProps<BookForm>) => {
  const navigate = useNavigate()
  return (
    <section>
      <Form>
        <Container fluid>
          <Row>
            <Col>
              {input({ name: 'title', placeholder: 'Catch-22' })}
              {input({ name: 'year', type: 'number' })}
              {input({
                name: 'description',
                type: 'textarea',
                placeholder: "That's some catch, that Catch-22",
                rows: '3',
              })}

              <Stars />

              <Field
                name="category"
                label="category"
                id="category"
                component={ReactstrapSelect}
                inputprops={{
                  name: 'category',
                  id: 'category',
                  options: categories,
                }}
              />

              <section>
                {submitError && <div>{submitError}</div>}
                <Button
                  onClick={() => navigate('/books')}
                  className="float-right "
                  color="secondary"
                >
                  Cancel
                </Button>
                <Button
                  className="float-right mr-2"
                  color="primary"
                  type="submit"
                  disabled={isSubmitting}
                >
                  Create
                </Button>
              </section>
            </Col>
          </Row>
        </Container>
      </Form>
    </section>
  )
}

export default NewBookWrapper
