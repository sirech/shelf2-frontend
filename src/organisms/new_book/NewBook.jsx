// @flow

import React from 'react'
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'

import { LinkContainer } from 'react-router-bootstrap'
import { Button, Container, Row, Col } from 'reactstrap'

import ReactstrapSelect from './ReactstrapSelect'
import input from './input'
import { Field, Form, Formik } from 'formik'

import initialBook from './initial_book'
import bookSchema from './schema'

import { actionPicker } from 'state'
import { actions } from 'state/form'

import type { Categories, BookForm as BookFormType } from 'types'
type Props = {
  history: Object,
  create: BookFormType => void,
}

const categories: Array<Categories> = [
  'sociology',
  'software',
  'econ',
  'history',
  'other',
]

const NewBookWrapper = (props: Props) => (
  <Formik
    initialValues={initialBook}
    initialStatus={{ submitError: '' }}
    render={NewBook}
    validationSchema={bookSchema}
    onSubmit={(values, { setStatus }) =>
      props.create(values, props.history, error =>
        setStatus({ submitError: error })
      )
    }
  />
)

const NewBook = ({ status: { submitError }, isSubmitting }: Object) => (
  <section>
    <Form>
      <Container fluid>
        <Row>
          <Col>
            {input({ name: 'title', placeholder: 'Catch-22' })}
            {input({ name: 'year', type: 'number' })}
            {input({
              name: 'description',
              type: 'textarea',
              placeholder: "That's some catch, that Catch-22",
              rows: '3',
            })}
            {input({ name: 'stars', type: 'number' })}

            <Field
              name="category"
              label="category"
              id="category"
              component={ReactstrapSelect}
              inputprops={{
                name: 'category',
                id: 'category',
                options: categories,
              }}
            />

            <section>
              {submitError && <div>{submitError}</div>}
              <LinkContainer to="/books">
                <Button className="float-right " color="secondary">
                  Cancel
                </Button>
              </LinkContainer>
              <Button
                className="float-right mr-2"
                color="primary"
                type="submit"
                disabled={isSubmitting}
              >
                Create
              </Button>
            </section>
          </Col>
        </Row>
      </Container>
    </Form>
  </section>
)

NewBookWrapper.defaultProps = {
  create: _ => undefined,
}

export default connect(
  null,
  actionPicker(['create'])(actions)
)(withRouter(NewBookWrapper))
