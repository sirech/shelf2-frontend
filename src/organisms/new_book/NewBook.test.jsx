import React from 'react'
import { fullRender } from 'test'
import NewBook from './NewBook'

describe('components', () => {
  describe('NewBook', () => {
    it('renders correctly', () => {
      const { component } = fullRender(<NewBook />)
      expect(component.toJSON()).toMatchSnapshot()
    })
  })
})
