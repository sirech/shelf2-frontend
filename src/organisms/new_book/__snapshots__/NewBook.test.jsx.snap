// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components NewBook renders correctly 1`] = `
<section>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="container-fluid"
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              className="label-color"
              htmlFor="title"
            >
              title
            </label>
            <input
              className="form-control"
              id="title"
              label="title"
              name="title"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Catch-22"
              type="text"
              value=""
            />
            
          </div>
          <div
            className="form-group"
          >
            <label
              className="label-color"
              htmlFor="year"
            >
              year
            </label>
            <input
              className="form-control"
              id="year"
              label="year"
              name="year"
              onBlur={[Function]}
              onChange={[Function]}
              type="number"
              value={2020}
            />
            
          </div>
          <div
            className="form-group"
          >
            <label
              className="label-color"
              htmlFor="description"
            >
              description
            </label>
            <textarea
              className="form-control"
              id="description"
              label="description"
              name="description"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="That's some catch, that Catch-22"
              rows="3"
              value=""
            />
            
          </div>
          <div
            className="form-group"
          >
            <label
              className="label-color"
              htmlFor="stars"
            >
              stars
            </label>
            <div>
              <span
                className="stars"
              >
                <span
                  className="star"
                  onClick={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-star fa-w-18 fa-lg "
                    data-icon="star"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </span>
                <span
                  className="star"
                  onClick={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-star fa-w-18 fa-lg "
                    data-icon="star"
                    data-prefix="far"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </span>
                <span
                  className="star"
                  onClick={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-star fa-w-18 fa-lg "
                    data-icon="star"
                    data-prefix="far"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </span>
                <span
                  className="star"
                  onClick={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-star fa-w-18 fa-lg "
                    data-icon="star"
                    data-prefix="far"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </span>
                <span
                  className="star"
                  onClick={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="svg-inline--fa fa-star fa-w-18 fa-lg "
                    data-icon="star"
                    data-prefix="far"
                    focusable="false"
                    role="img"
                    style={Object {}}
                    viewBox="0 0 576 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
                      fill="currentColor"
                      style={Object {}}
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
          <div
            className="form-group"
          >
            <label
              className="label-color"
              htmlFor="category"
            >
              category
            </label>
            <select
              className="form-control"
              id="category"
              inputprops={
                Object {
                  "id": "category",
                  "name": "category",
                  "options": Array [
                    "sociology",
                    "software",
                    "econ",
                    "history",
                    "other",
                  ],
                }
              }
              label="category"
              name="category"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Test"
              value="software"
            >
              <option
                value=""
              />
              <option
                value="sociology"
              >
                sociology
              </option>
              <option
                value="software"
              >
                software
              </option>
              <option
                value="econ"
              >
                econ
              </option>
              <option
                value="history"
              >
                history
              </option>
              <option
                value="other"
              >
                other
              </option>
            </select>
          </div>
          <section>
            
            <a
              aria-label={null}
              className="float-right btn btn-secondary"
              href="/books"
              onClick={[Function]}
            >
              Cancel
            </a>
            <button
              aria-label={null}
              className="float-right mr-2 btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              Create
            </button>
          </section>
        </div>
      </div>
    </div>
  </form>
</section>
`;
